README.md
This forks runs the _Betapose training_ on Windows 10 with CUDA 10.1

#	Betapose: Estimating 6D Pose From Localizing Designated Surface Keypoints
Please refer to our paper for detailed explanation. Arxiv Link is [here](https://arxiv.org/abs/1812.01387).
In the following, `ROOT` refers to the folder containing this README file.
## 0. News:
   I contained all important public literature in this repo to boost research. If there is a conflict or an issue, please raise it in issue part.
## Ⅰ. Installation
1. All the codes are tested in Python, CUDA 8.0 and CUDNN 5.1.
2. Install [pytorch 0.4.0](https://github.com/pytorch/pytorch) and other dependencies.
3. Download LineMod dataset [here](http://ptak.felk.cvut.cz/6DB/public/datasets/hinterstoisser/). Only folders called models and test are needed. Put them in `DATAROOT/models` and `DATAROOT/test` where `DATAROOT` can be any folder you'd like to place LineMod dataset.
## Ⅱ. Designate Keypoints
**You can skip this step since we have provided designated keypoints files in '$ROOT/1_keypoint_designator/assets/sifts/'**.
1. The related code is in `$ROOT/1_keypoint_designator/`.
	```bash
	$ cd ROOT/1_keypoint_designator/
    ```
 2. Place the input model file (e.g. `DATAROOT/models/obj_01.ply`) in `$./assets/models/` 
 3. Build the code and run it. Just type:
 	```bash
	$ sh build_and_run.sh
	```
	The output file is in `$./assets/sifts/`. It's a ply file storing the 3D coordinates of designated keypoints.
## Ⅲ. Annotate Keypoints
1. The related code is in `$ROOT/2_keypoint_annotator/`.
	```bash
	$ cd ROOT/2_keypoint_annotator/
	```
2. Run keypoint annotator on one object of LineMod.
	```bash
	$ python annotate_keypoint.py --obj_id 1 --total_kp_number 50 --output_base ROOT/3_6Dpose_estimator/data --sixd_base DATAROOT
	```
	Type the following to see the meaning of options.
	```bash
	$ python annotate_keypoint.py -h
	```	
3. The annotated keypoints are in file `annot_train.h5` and `annot_eval.h5`. The corresponding training images are in folders `train` and `eval`.
## Ⅳ. Training
### Train Object Detector YOLOv3
1. Relative files locate in `$ROOT/3_6Dpose_estimator/train_YOLO`.
	```bash
	$ cd ROOT/3_6Dpose_estimator/train_YOLO
	```
2. Build Darknet (YOLOv3).
	```bash
	$ make
	```	
3. Prepare data as [AlexeyAB/darknet](https://github.com/AlexeyAB/darknet)'s instructions. Refer to folder `./scripts` for more help.
4. Download pretrained darknet53 [here](http://pjreddie.com/media/files/darknet53.conv.74)
5. Run `train_single.sh` or `train_all.sh` to train the network. 
6. Put trained weights (e.g. 01.weights) in folder `$ROOT/3_6Dpose_estimator/models/yolo/`.
### Train Keypoint Detector (KPD)
1.  Relative code is in `$ROOT/3_6Dpose_estimator/train_KPD/`
	```bash
	$ cd ROOT/3_6Dpose_estimator/train_KPD
	```
3. Modify Line 19, 21, 39, 46 of file`./src/utils/dataset/coco.py` to previously annotated dataset. Examples are given in these lines.
2. **Train on Linemod dataset without DPG**. 
	```bash
	$ python src/train.py --trainBatch 28 --expID seq5_Nov_1_1 --optMethod adam
	```
3. **Train on Linemod dataset with DPG**.  Just add a `--addDPG` option. and load the model trained after in the second step.
	```bash
	$ python src/train.py --trainBatch 28 --expID seq5_dpg_Nov_1_1 --optMethod adam --loadModel ./exp/coco/seq5_Nov_1_1/model_100.pkl --addDPG
	```
5. (Optional) Visualize training process. Type
	```bash
	$ tensorboard --logdir ./
	```

## Ⅴ. Evaluate
1.  Move back to the root of pose estimator.
	```bash
	$ cd ROOT/3_6Dpose_estimator/
	``` 
2. Run the following command.

	```bash
	$ CUDA_VISIBLE_DEVICES=1 python3 betapose_evaluate.py --nClasses 50 --indir /01/eval --outdir examples/seq1 --sp --profile
	```
	The output json file containing predicted 6D poses will be in examples/seq1.


CHANGELOG.md
# Changelog

* Updated code to run on Python 3.7, PyTorch 1.0.1, CUDA 10 and CUDNN 7.5 (_tested on Windows_)

* (TODO) Added missing keypoints files for class 3 and 7

* Added batch scripts (for automating tasks on Windows):
  * `annotate_all.bat`: automate annotation tasks (II) for each LineMod class
  * `train_single.bat`: launch training on darknet for the designated class (modify batch file to change class)

* Modified `yolo-linemod-single.cfg`:
  * Set batch_size to 32 (instead of 64) because of hardware limitation
  * Limited max_batches to 10000 (instead of 500200) for easier training time management

* **(TODO)** Added pre-trained weights for each class for quick evaluation

* **(TODO)** Updated `README.md`


3_6Dpose_estimator/opt.py
import argparse
import torch

parser = argparse.ArgumentParser(description='PyTorch AlphaPose Training')

"----------------------------- General options -----------------------------"
parser.add_argument('--expID', default='default', type=str,
                    help='Experiment ID')
parser.add_argument('--dataset', default='coco', type=str,
                    help='Dataset choice: mpii | coco')
parser.add_argument('--nThreads', default=40, type=int,
                    help='Number of data loading threads')
parser.add_argument('--left_keypoints', default=10, type=int,
                    help='Number of keypoints left in PnP algorithm.')
parser.add_argument('--debug', default=False, type=bool,
                    help='Print the debug information')
parser.add_argument('--snapshot', default=1, type=int,
                    help='How often to take a snapshot of the model (0 = never)')
parser.add_argument('--obj_id', dest="object_id", default=5, type=int,
                    help='Object ID to test!')
parser.add_argument('--img_id', dest="image_id", default=5, type=int,
                    help='Image ID to test!')

"----------------------------- AlphaPose options -----------------------------"
parser.add_argument('--addDPG', default=False, type=bool,
                    help='Train with data augmentation')
parser.add_argument('--sp', default=False, action='store_true',
                    help='Use single process for pytorch')
parser.add_argument('--profile', default=False, action='store_true',
                    help='add speed profiling at screen output')

"----------------------------- Model options -----------------------------"
parser.add_argument('--netType', default='hgPRM', type=str,
                    help='Options: hgPRM | resnext')
parser.add_argument('--loadModel', default=None, type=str,
                    help='Provide full path to a previously trained model')
parser.add_argument('--Continue', default=False, type=bool,
                    help='Pick up where an experiment left off')
parser.add_argument('--nFeats', default=256, type=int,
                    help='Number of features in the hourglass')
parser.add_argument('--nClasses', default=50, type=int,
                    help='Number of output channel')
parser.add_argument('--nStack', default=4, type=int,
                    help='Number of hourglasses to stack')

"----------------------------- Hyperparameter options -----------------------------"
parser.add_argument('--fast_inference', default=True, type=bool,
                    help='Fast inference')
parser.add_argument('--use_pyranet', default=True, type=bool,
                    help='use pyranet')

"----------------------------- Hyperparameter options -----------------------------"
parser.add_argument('--LR', default=2.5e-4, type=float,
                    help='Learning rate')
parser.add_argument('--momentum', default=0, type=float,
                    help='Momentum')
parser.add_argument('--weightDecay', default=0, type=float,
                    help='Weight decay')
parser.add_argument('--crit', default='MSE', type=str,
                    help='Criterion type')
parser.add_argument('--optMethod', default='rmsprop', type=str,
                    help='Optimization method: rmsprop | sgd | nag | adadelta')


"----------------------------- Training options -----------------------------"
parser.add_argument('--nEpochs', default=200, type=int,
                    help='Number of hourglasses to stack')
parser.add_argument('--epoch', default=0, type=int,
                    help='Current epoch')
parser.add_argument('--trainBatch', default=40, type=int,
                    help='Train-batch size')
parser.add_argument('--validBatch', default=20, type=int,
                    help='Valid-batch size')
parser.add_argument('--trainIters', default=0, type=int,
                    help='Total train iters')
parser.add_argument('--valIters', default=0, type=int,
                    help='Total valid iters')
parser.add_argument('--init', default=None, type=str,
                    help='Initialization')

"----------------------------- Data options -----------------------------"
parser.add_argument('--inputResH', default=320, type=int,
                    help='Input image height')
parser.add_argument('--inputResW', default=256, type=int,
                    help='Input image width')
parser.add_argument('--outputResH', default=80, type=int,
                    help='Output heatmap height')
parser.add_argument('--outputResW', default=64, type=int,
                    help='Output heatmap width')
parser.add_argument('--scale', default=0.25, type=float,
                    help='Degree of scale augmentation')
parser.add_argument('--rotate', default=30, type=float,
                    help='Degree of rotation augmentation')
parser.add_argument('--hmGauss', default=1, type=int,
                    help='Heatmap gaussian size')

"----------------------------- PyraNet options -----------------------------"
parser.add_argument('--baseWidth', default=9, type=int,
                    help='Heatmap gaussian size')
parser.add_argument('--cardinality', default=5, type=int,
                    help='Heatmap gaussian size')
parser.add_argument('--nResidual', default=1, type=int,
                    help='Number of residual modules at each location in the pyranet')

"----------------------------- Distribution options -----------------------------"
parser.add_argument('--dist', dest='dist', type=int, default=1,
                    help='distributed training or not')
parser.add_argument('--backend', dest='backend', type=str, default='gloo',
                    help='backend for distributed training')
parser.add_argument('--port', dest='port',
                    help='port of server')

"----------------------------- Detection options -----------------------------"
parser.add_argument('--net', dest='demo_net', help='Network to use [vgg16 res101]',
                    default='res152')
parser.add_argument('--indir', dest='inputpath',
                    help='image-directory', default="")
parser.add_argument('--list', dest='inputlist',
                    help='image-list', default="")
parser.add_argument('--mode', dest='mode',
                    help='detection mode, fast/normal/accurate', default="normal")
parser.add_argument('--outdir', dest='outputpath',
                    help='output-directory', default="examples/res/")
parser.add_argument('--inp_dim', dest='inp_dim', type=str, default='416',
                    help='inpdim')
parser.add_argument('--conf', dest='confidence', type=float, default=0.01,
                    help='bounding box confidence threshold')
parser.add_argument('--nms', dest='nms_thesh', type=float, default=0.6,
                    help='bounding box nms threshold')
parser.add_argument('--save_img', default=False, action='store_true',
                    help='save result as image')
parser.add_argument('--vis', default=False, action='store_true',
                    help='visualize image')
parser.add_argument('--format', type=str,
                    help='save in the format of cmu or coco or openpose, option: coco/cmu/open')
parser.add_argument('--detbatch', type=int, default=1,
                    help='detection batch size')
parser.add_argument('--posebatch', type=int, default=80,
                    help='pose estimation maximum batch size')

"----------------------------- Video options -----------------------------"
parser.add_argument('--video', dest='video',
                    help='video-name', default="")
parser.add_argument('--webcam', dest='webcam', type=str,
                    help='webcam number', default='0')
parser.add_argument('--save_video', dest='save_video',
                    help='whether to save rendered video', default=False, action='store_true')
parser.add_argument('--vis_fast', dest='vis_fast',
                    help='use fast rendering', action='store_true', default=False)
opt = parser.parse_args()

opt.num_classes = 80
